### PR TITLE
Merging function and TemplatedFunction

### PR DESCRIPTION
**Pull request by Oshanath**
reference tests pass. I checked and as far as I saw only the extra [] for typevars are added to them.

`ctest` still dont pass. I will change them next. I will manually generate the ASR for them and paste the result. I checked and there too, the only difference is the extra [] for the typevars.

### CODE REVIEW

**Comment by certik** [comment]
Thanks @Oshanath.  Here is a TODO:

* [ ] Merge the latest `main` into this branch. The `main` now has the code to represent and instantiate templates
* [ ] change the `swap` test to an `add` test
* [ ] improve the code to load the new test to ASR, see if we can exactly reuse the existing ASR from `main`
* [ ] At the instantiation site, simply call the instantiate code, just like LPython does

Ping me if you run into any issues or have any questions.

**Comment by Oshanath** [comment]
In the instantiation, I get the types as `decl_attribute`s. I need to convert this into `ttype`s. For that currently I'm thinking I should use the same code from [here](https://github.com/lfortran/lfortran/blob/d5fc7145897d3d1c2a9bf2adadc6a792cbf1fc43/src/lfortran/semantics/ast_common_visitor.h#L923). Am I correct?

Shall I refactor that part into a new function so that I can use it? Or is there an easier way?

**Comment by certik** [comment]
Yes, I think you are correct.

That part in AST->ASR in LFortran is quite complex already, so we should figure out how to refactor it. For now just do what you can, refactor it for your needs. We'll come back to it later.

**Comment by Oshanath** [comment]
@certik  Users call the instantiation by the name of the template. But currently, the name of the template is not stored anywhere. So I think we should either add a new parameter to Functions to store to which template that function belongs to. Or we should create an ASR node called Template to store its name and the type parameters. I think the latter is the better choice. What do you think?

**Comment by certik** [comment]
I see. I need to think about this, also with relation to restrictions, see Zulip.

**Comment by Oshanath** [comment]
@certik I had change the naming process of the instantiation because the new function name is determined by the user.

File: examples/add_m.f90 [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+contains
+
+    subroutine test_template()
+        instantiate add_t(real), only: add_real => add_generic
+        real :: x, y
+        integer :: a, b
+        x = 5.1
+        y = 7.2
+        print*, "The result is ", add_real(x, y)
+
```
Comment (certik) [inline examples/add_m.f90:31]:
```suggestion
if (abs(add_real(x, y) - 12.3) > 1e-5) error stop

```

File: examples/add_m.f90 [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        instantiate add_t(integer), only: add_integer => add_generic
+        a = 5
+        b = 9
+        print*, "The result is ", add_integer(a, b)
+
```
Comment (certik) [inline examples/add_m.f90:36]:
```suggestion
if (add_integer(x, y) /= 14) error stop
```

File: tests/reference/asr-swap_m-c05f43c.stderr [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    case unit_decl2Type::Declaration: { v.visit_Declaration((const Declaration_t &)x); return; }
+[2m  File "[0m[1m[35m$DIR/src/lfortran/semantics/ast_symboltable_visitor.cpp[39m[0m[2m", line 572, in LFortran::SymbolTableVisitor::visit_Declaration(LFortran::AST::Declaration_t const&)[0m
+    visit_DeclarationUtil(x);
+[2m  File "[0m[1m[35m$DIR/src/lfortran/semantics/ast_common_visitor.h[39m[0m[2m", line 1086, in LFortran::CommonVisitor<LFortran::SymbolTableVisitor>::visit_DeclarationUtil(LFortran::AST::Declaration_t const&)[0m
+    if( is_c_ptr(v, derived_type_name) ) {
+[2m  File "[0m[1m[35m$DIR/src/lfortran/semantics/ast_common_visitor.h[39m[0m[2m", line 728, in LFortran::CommonVisitor<LFortran::SymbolTableVisitor>::is_c_ptr(LFortran::ASR::symbol_t*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)[0m
+    ASR::symbol_t* v_orig = ASRUtils::symbol_get_past_external(v);
+[2m  File "[0m[1m[35m$DIR/src/libasr/asr_utils.h[39m[0m[2m", line 43, in ??[0m
+    if (f->type == ASR::symbolType::ExternalSymbol) {
+[2m  Binary file "[0m[1m[35m/lib/x86_64-linux-gnu/libc.so.6[39m[0m[2m", in killpg()[0m
```
Comment (certik) [inline tests/reference/asr-swap_m-c05f43c.stderr:55]: can be removed now.

File: tests/tests.toml [hunk 4]
```diff
@@ -11,6 +11,11 @@ ast_indent = true
 asr_indent = true
 ast = true
 
+[[test]]
+filename = "../examples/add_m.f90"
```
Comment (certik) [inline tests/tests.toml:15]: Move this test from examples into integration_tests.

**Comment by certik** [comment]
This is the latest PR, and it works!!!

So all we need is to just polish this up and merge into master.

**Comment by Oshanath** [comment]
@certik I applied your suggestions, moved the example to integration tests and cleaned up the history.

**Comment by certik** [comment]
I went over all commits, everything looks good.

Thanks for finishing this up!

**Pull request merged**